File: Tree/1110-deleteNodes.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

'''
Given the root of a binary tree, each node in the tree has a distinct value.

After deleting all nodes with a value in to_delete, we are left with a forest (a disjoint union of trees).

Return the roots of the trees in the remaining forest. You may return the result in any order.
Input: root = [1,2,3,4,5,6,7], to_delete = [3,5]
Output: [[1,2,null,4],[6],[7]]

'''
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def delNodes(self, root, to_delete):
        if not root:
            return []
        ans=[]
        ds=set(to_delete)
        def process(root):
            if not root:
                return None
            root.left=process(root.left)
            root.right=process(root.right)
            if root.val not in ds:
                return root
            if root.left:
                ans.append(root.left.val)
            if root.right:
                ans.append(root.right.val)
        
        root=process(root)
        if root:
            ans.append(root.val)
        return ans

root=TreeNode(1)
two=TreeNode(2)
three=TreeNode(3)
four=TreeNode(4)
five=TreeNode(5)
root.left=two
root.right=three

two.left=four
three.right=five
to_delete=[2,3]
print(Solution().delNodes(root, to_delete))

File: Tree/1110-deleteNodes.cpp
#include "treeNode.h"
#include<vector>
#include<iostream>
#include<algorithm> //for std::find
using VEC=std::vector<int>;
using std::cout;

TreeNode* process(TreeNode* root, VEC& to_del, VEC& ans)
{
    if(!root)
        return nullptr;
    root->left=process(root->left,to_del,ans);
    root->right=process(root->right,to_del,ans);
    if(std::find(to_del.begin(),to_del.end(),root->m_val)==to_del.end())
        return root;
    if(root->left)
        ans.push_back(root->left->m_val);
    if(root->right)
        ans.push_back(root->right->m_val);
    return nullptr;
}


VEC delNodes(TreeNode* root, VEC& to_delete)
{
    VEC ans{};
    if(!root)
        return ans;
    root=process(root,to_delete,ans);
    if(root)
        ans.push_back(root->m_val);
    return ans;
}

int main()
{
        TreeNode *root=new TreeNode(1);
    TreeNode *two=new TreeNode(2);
    TreeNode *three=new TreeNode(3);
    TreeNode *four=new TreeNode(4);
    TreeNode *five=new TreeNode(5);
    // TreeNode *six=new TreeNode(6);
    // TreeNode *seven=new TreeNode(7);

    root->left=two;
    root->right=three;

    two->left=four;
    three->right=five;
    VEC to_del{2,3};
    VEC ans=delNodes(root,to_del);
    for(auto i:ans)
        cout<<i<<',';
}